### DISCUSSION TITLE
Visit provider at a position without UI

### DISCUSSION THREAD

**Discussion by lizhuoyuan** [rrousselGit/riverpod#2215] [Q&A]
First express my gratitude, then please allow me to describe my needs simply.
I have a provider:

```
class MessageNotifier extends StateNotifier<List<RCIMIWConversation>> {
  MessageNotifier() : super([]);

  void update(List<RCIMIWConversation> value) {
    state = value;
  }
}

final messageProvider =
    StateNotifierProvider<MessageNotifier, List<RCIMIWConversation>>(
        (ref) => MessageNotifier());

```
I have a util:
```
class MyUtil {
...
 static Future<void> init() async {
...
//this is a listener
 listen((value){
  //i want to visit other provider ,e.g.
 ref.read(messageProvider.notifer).update(value);
});
}
}

```
Because there is no Ref here, I actually can't do this. What is the best way I want to do?
 
Myutil is a class with several static methods.
Maybe a ref parameter should be passed in?
Maybe I should modify it to provider? I'm not sure.
Maybe my question is, where there is no UI, or there is no REF, what should I do if I want to use Provider?

**Comment by AhmedLSayed9** [comment]
> where there is no UI, or there is no REF, what should I do if I want to use Provider?

You should access provider using `ref`. If you don't have one, then make a provider for your `MyUtil` class and pass ref to its constructor.

**Reply by lizhuoyuan** [reply]
All the methods in MyUtil are static, so I have to change the method to non-static, and change the calling method to call through its instance, right?

**Reply by AhmedLSayed9** [reply]
A static method can't access class's non-static members (i.e: ref).
You might need to share your full example as I'm not sure how/what the listener inside init will be doing.

**Reply by lizhuoyuan** [reply]
this is myutil code:
```
class MyUtil {
  static const appKey = 'appkey';
  static RCIMIWEngine? engine;

  static Future<void> init({isRefresh = false}) async {
  
    ....
    engine ??= await RCIMIWEngine.create(appKey, options);

    engine?.onConversationsLoaded = (int? code,
        List<RCIMIWConversationType>? conversationTypes,
        String? channelId,
        int? startTime,
        int? count,
        List<RCIMIWConversation>? conversations) {
      debugPrint('[RongCloud] onConversationsLoaded: $conversations');

      //here want to use provider
      mainRef.read(messageProvider.notifier).update(conversations ?? []);
     
    };
  }

  static Future sendTextMessage(
    
  }) async {
    
  }

  static Future logout({bool receivePush = false}) async {
  
  }

  static Future dispose() async {
     
  }

  static Future<int> totalUnreadCount() async {
   
  }
```

**Comment by ninhngadev** [comment]
You can use ProviderContainer.

**Reply by AhmedLSayed9** [reply]
That's anti-pattern.

**Comment by lizhuoyuan** [comment]
I have a simple way to define a global mainRef in main.dart so that I can get the ref elsewhere, but I'm not sure if there are potential problems with this approach

**Reply by AhmedLSayed9** [reply]
Again, It's an anti-pattern. You don't want to expose your container globally.
